### PR TITLE
Replace boolean gui.IsInMessageBox with function gui.message.isModalMessageBoxActive

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:

Relates to #13007, as the new design should be backwards compatible, we can implement the API changes as-is for 2022.1.
Follow up to the API proposed in #12984

### Summary of the issue:

The API for `gui.message` had not yet been determined for 2022.1, as per #13007.
As the future API is intended to be backwards compatible, this PR commits to the API proposed in #12984.

### Description of how this pull request fixes the issue:

`gui.IsInMessageBox` was a boolean, this has been replaced by a function `gui.message.isModalMessageBoxActive`, which tracks if a messageBox is open in a safer manner.

Changes logic `gui/message.py` to be safer with locking and handling errors.

### Testing strategy:
Alpha/beta testing of messageBoxes 

### Known issues with pull request:
None

### Change log entries:
See PR diff

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

File: user_docs/en/changes.t2t [hunk 1]
```diff
@@ -133,6 +133,7 @@ This can dramatically decrease build times on multi core systems. (#13226)
   - ``cachedAutomationId`` can be used if obtained directly from the element.
   -
 - ``NVDAObjects.window.scintilla.CharacterRangeStruct`` has moved to ``NVDAObjects.window.scintilla.Scintilla.CharacterRangeStruct``. (#13364)
+- ``gui.isInMessageBox`` is now a function which returns a boolean. (#12984, #13376)
```
Comment (feerrenrut) [inline user_docs/en/changes.t2t:136]: What was it before?

**Review by feerrenrut — CHANGES_REQUESTED** [review]
if `isInMessageBox` is still used as a bool variable, it will always be true. I think a new name should be used, so that that code that doesn't get updated throws an error, rather than doing the wrong thing. Consider:

``` python
def isInMessageBox():
	return False
print(bool(isInMessageBox))  # prints True

# contrived example of addon code that has not been updated.
if isInMessageBox:  # always true, instead it should error "name 'isInMessageBox' is not defined"
    log.debug("in the message box")
```

**Comment by feerrenrut** [comment]
Could you update this PR to make the title more descriptive of the change, and describe the changes that are part of this PR. Issue https://github.com/nvaccess/nvda/pull/12984 contains lots of suggestions, what does this PR do?

File: user_docs/en/changes.t2t [hunk 2]
```diff
@@ -131,6 +131,7 @@ This can dramatically decrease build times on multi core systems. (#13226)
   - ``cachedAutomationId`` can be used if obtained directly from the element.
   -
 - ``NVDAObjects.window.scintilla.CharacterRangeStruct`` has moved to ``NVDAObjects.window.scintilla.Scintilla.CharacterRangeStruct``. (#13364)
+- ``gui.isInMessageBox`` is removed, please use ``gui.message.isMessageBoxActive`` instead. (#12984, #13376)
```
Comment (feerrenrut) [inline user_docs/en/changes.t2t:134]:
```suggestion
- Boolean ``gui.isInMessageBox`` is removed, please use function ``gui.message.isMessageBoxActive()`` instead. (#12984, #13376)

```

**Review by feerrenrut — COMMENTED** [review]
Don't forget to update the API plan. 

Thinking about what is important to point out in the plan:
- This function can only tell you if 1 or more message boxes are active.
- If, in the future, there can be multiple (non-modal) message boxes active, the API will need to be extended, perhaps with "getActiveMesageBoxes()". This is fine.

This second point makes me think this method should be specific to modal message boxes, I.E. rename to `isModalMessageBoxActive()`. With docs explicitly stating "only one modal message box may be active at once"

What do you think?

**Comment by seanbudd** [comment]
>With docs explicitly stating "only one modal message box may be active at once"

I don't believe this is true, as per https://github.com/nvaccess/nvda/pull/12984#issuecomment-952561745.

I think the name `isModalMessageBoxBlocking` might even be better. In the future API we might want to ignore non-blocking message boxes.

**Comment by feerrenrut** [comment]
Yes, I'd forgotten that case, thought it is arguable that even in that case only one is active at a time. I.E. only one of them is processing messages.

> I think the name isModalMessageBoxBlocking might even be better.

It raises the question, what is it blocking?

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/gui/message.py [hunk 3]
```diff
@@ -25,30 +39,38 @@ def messageBox(
...
-	This should be used for all message dialogs
-	rather than using C{wx.MessageDialog} and C{wx.MessageBox} directly.
+	Avoid using C{wx.MessageDialog} and C{wx.MessageBox} directly.
 	@param message: The message text.
 	@param caption: The caption (title) of the dialog.
 	@param style: Same as for wx.MessageBox.
 	@param parent: The parent window.
 	@return: Same as for wx.MessageBox.
+
+	messageBox is a function which blocks the calling thread,
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/gui/message.py]: fixed, by changing this to `gui.message.messageBox`

File: source/gui/message.py [hunk 4]
```diff
@@ -14,8 +13,23 @@
 _messageBoxCounter = 0
 
 
-def isInMessageBox() -> bool:
-	return _messageBoxCounter != 0
+def isModalMessageBoxActive() -> bool:
+	"""
+	messageBox is a function which blocks the calling thread,
+	until a user responds to the modal dialog.
+	`IsInMessageBox` should be used when an answer is required before proceeding.
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/gui/message.py]: This should actually refer to `isModalMessageBoxActive`, which has been fixed and the sentence reworded to be clearer.

```
`gui.message.messageBox` is a function which blocks the calling thread,
until a user responds to the modal dialog.
`isModalMessageBoxActive` should be used when an answer from a `gui.message.messageBox`
is required before proceeding.
```

File: source/gui/message.py [hunk 5]
```diff
@@ -14,8 +13,23 @@
...
-	return _messageBoxCounter != 0
+def isModalMessageBoxActive() -> bool:
+	"""
+	messageBox is a function which blocks the calling thread,
+	until a user responds to the modal dialog.
+	`IsInMessageBox` should be used when an answer is required before proceeding.
+	NVDA is in an uncertain state while waiting for an answer.
+
+	It's possible for multiple message boxes to be open at a time.
+	This function may prevent this from happening.
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/gui/message.py]: fixed

File: source/gui/message.py [hunk 6]
```diff
@@ -25,30 +39,38 @@ def messageBox(
...
 	@param message: The message text.
 	@param caption: The caption (title) of the dialog.
 	@param style: Same as for wx.MessageBox.
 	@param parent: The parent window.
 	@return: Same as for wx.MessageBox.
+
+	messageBox is a function which blocks the calling thread,
+	until a user responds to the modal dialog.
+	This function should be used when an answer is required before proceeding.
+	Consider using a custom subclass of a wxDialog if an answer is not required.
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/gui/message.py]: added

File: source/gui/message.py [hunk 7]
```diff
@@ -25,30 +39,38 @@ def messageBox(
...
+
+	It's possible for multiple message boxes to be open at a time.
+	Before opening a new messageBox, use isModalMessageBoxResponsePending
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/gui/message.py]: fixed

File: source/gui/message.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/gui/message.py]: > `isModalMessageBoxActive` should be used when an answer from a `gui.message.messageBox`
is required before proceeding.

I don't think it's obvious why you would call `isModalMessageBoxActive` if an answer is required. You'd call it to ensure there isn't already a modal message box open.

[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/gui/message.py]: The current sentence makes it sound like calling `isModalMessageBoxActive` makes the answer required.

[the remark this answers is not in the document]
Comment (seanbudd) [reply source/gui/message.py]: >I don't think it's obvious why you would call isModalMessageBoxActive if an answer is required. You'd call it to ensure there isn't already a modal message box open.

These two things are the same, no? This doesn't answer the question as to why you should ensure at modal message box isn't currently open. To safely proceed with shutdown, you need an answer from pending messageBoxes. To safely proceed with opening certain dialogs and other messageboxes, you need an answer from pending messageBoxes.

How about:

>`isModalMessageBoxActive` should be used when an answer from an open `gui.message.messageBox`
is required before proceeding.

[the remark this answers is not in the document]
Comment (seanbudd) [reply source/gui/message.py]: Without this sentence, I think it is unclear on the purpose of this function, and when to use it.

[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/gui/message.py]: It still make it sound like it is `isModalMessageBoxActive` that makes the answer a requirement.

Instead:
```
When some action (E.G. Quitting NVDA) should be prevented due to any active modal/blocking message box,
even if unrelated, use `isModalMessageBoxActive` to check before triggering the action, 
```

[the remark this answers is not in the document]
Comment (seanbudd) [reply source/gui/message.py]: amended in 8f28274

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**